### PR TITLE
Encrypted handshake

### PR DESCRIPTION
**Pull request by rabbitonweb**
Fixes CORE-171

### CODE REVIEW

**Comment by rabbitonweb** [comment]
@kirkwood @echatav this is still work in progress (however close to be done) but since a lot of that stuff is also refactoring (introducing monadic api), i would be happy if we review what is here so far. 

If (worst case scenario) this whole thing goes to trash, I don't want to lose changes I did while hacking the protocol. Thus kindly asking for review of current state.

File: comm/src/main/scala/coop/rchain/catscontrib/Capture.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  */
+trait Capture[F[_]] {
+
+  /** Captures the effect of producing `A`, including any exceptions that may
+    * be thrown.
+    */
+  def capture[A](a: => A): F[A]
+
+  /** Alias for `capture`. */
+  def apply[A](a: => A): F[A] = capture(a)
```
Comment (henrytill) [inline comm/src/main/scala/coop/rchain/catscontrib/Capture.scala]: Double thunk?

**Comment by rabbitonweb** [comment]
@henrytill I'm slowly building `catscontrib` where I will be adding stuff that are missing - https://github.com/rchain/rchain/pull/279/files#diff-338fabe1f3dd3c0baf4592ec73e4a946. We could think about creating a separate module for it I guess

File: comm/src/main/scala/coop/rchain/catscontrib/Capture.scala (hunk 1 above)
Comment (rabbitonweb) [reply comm/src/main/scala/coop/rchain/catscontrib/Capture.scala]: yeah, will fix that, thx!

File: comm/src/main/scala/coop/rchain/catscontrib/Capture.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package coop.rchain.catscontrib
```
Comment (henrytill) [inline comm/src/main/scala/coop/rchain/catscontrib/Capture.scala]: potentially we could start another top-level project for this.  i have some code that go there as well...

File: comm/src/main/scala/coop/rchain/catscontrib/Capture.scala (hunk 1 above)
Comment (rabbitonweb) [reply comm/src/main/scala/coop/rchain/catscontrib/Capture.scala]: and.... it's gone

File: comm/src/main/scala/coop/rchain/catscontrib/Capture.scala (hunk 2 above)
Comment (henrytill) [reply comm/src/main/scala/coop/rchain/catscontrib/Capture.scala]: perhaps `library` so it's more general than cats extension stuff?

Comment (rabbitonweb) [reply comm/src/main/scala/coop/rchain/catscontrib/Capture.scala]: reading my mind https://github.com/rchain/rchain/pull/279#issuecomment-366083133 :)

File: comm/src/main/protobuf/coop/rchain/comm/protocol/p2p.proto [hunk 3]
```diff
@@ -16,11 +16,13 @@ message Hello {
 }
 
 message EncryptionHandshake {
```
Comment (rabbitonweb) [inline comm/src/main/protobuf/coop/rchain/comm/protocol/p2p.proto]: rename it, to sth like IdentificationHandshake

**Comment by echatav** [comment]
Just make sure it passes CI before merging!

**Pull request merged**